L1Trigger/L1CaloTrigger/plugins/Phase1L1TJetProducer.cc

```diff
@@ -0,0 +1,339 @@
+  std::vector<reco::CaloJet> l1jetVector;
+  if (puSubtraction_) {
+    l1jetVector = _buildJetsFromSeedsWithPUSubtraction(*(caloGrid_), seedsVector, vetoZeroPt_);
+  } else {
+    l1jetVector = _buildJetsFromSeeds(*(caloGrid_), seedsVector);
+  }
+
+  // sort by pt
+  std::sort(l1jetVector.begin(), l1jetVector.end(), [](const reco::CaloJet& jet1, const reco::CaloJet& jet2) {
+    return jet1.pt() > jet2.pt();
+  });
+
+  std::unique_ptr<std::vector<reco::CaloJet>> l1jetVectorPtr(new std::vector<reco::CaloJet>(l1jetVector));
+  iEvent.put(std::move(l1jetVectorPtr), outputCollectionName_);
+
+  return;
+}
+
+void Phase1L1TJetProducer::subtract9x9Pileup(const TH2F& caloGrid, reco::CaloJet& jet) {
+  // these variables host the total pt in each sideband and the total pileup contribution
+  float topBandPt = 0;
+  float leftBandPt = 0;
+  float rightBandPt = 0;
+  float bottomBandPt = 0;
+  float pileUpEnergy;
+
+  // hold the jet's x-y (and z, as I have to use it, even if 2D) location in the histo
+  int xCenter, yCenter, zCenter;
+  // Retrieving histo-coords for seed
+  caloGrid.GetBinXYZ(caloGrid.FindFixBin(jet.eta(), jet.phi()), xCenter, yCenter, zCenter);
+
+  // Computing pileup
+  for (int x = -4; x <= 4; x++) {
+    for (int y = 0; y < 3; y++) {
+      // top band, I go up 5 squares to reach the bottom of the top band
+      // +x scrolls from left to right, +y scrolls up
+      topBandPt += getTowerEnergy(caloGrid, xCenter + x, yCenter + (5 + y));
+      // left band, I go left 5 squares (-5) to reach the bottom of the top band
+      // +x scrolls from bottom to top, +y scrolls left
+      leftBandPt += getTowerEnergy(caloGrid, xCenter - (5 + y), yCenter + x);
+      // right band, I go right 5 squares (+5) to reach the bottom of the top band
+      // +x scrolls from bottom to top, +y scrolls right
+      rightBandPt += getTowerEnergy(caloGrid, xCenter + (5 + y), yCenter + x);
+      // right band, I go right 5 squares (+5) to reach the bottom of the top band
+      // +x scrolls from bottom to top, +y scrolls right
+      bottomBandPt += getTowerEnergy(caloGrid, xCenter + x, yCenter - (5 + y));
+    }
+  }
+  // adding bands and removing the maximum band (equivalent to adding the three minimum bands)
+  pileUpEnergy = topBandPt + leftBandPt + rightBandPt + bottomBandPt -
+                 std::max(topBandPt, std::max(leftBandPt, std::max(rightBandPt, bottomBandPt)));
+
+  //preparing the new 4-momentum vector
+  reco::Candidate::PolarLorentzVector ptVector;
+  // removing pu contribution
+  float ptAfterPUSubtraction = jet.pt() - pileUpEnergy;
+  ptVector.SetPt((ptAfterPUSubtraction > 0) ? ptAfterPUSubtraction : 0);
+  ptVector.SetEta(jet.eta());
+  ptVector.SetPhi(jet.phi());
+  //updating the jet
+  jet.setP4(ptVector);
+  jet.setPileup(pileUpEnergy);
+  return;
+}
+
+std::vector<std::tuple<int, int>> Phase1L1TJetProducer::findSeeds(const TH2F& caloGrid, float seedThreshold) {
```

Comment:
function can be `const`
this one and the above could also just use the `caloGrid_` member variable directly

Reply:
done: https://github.com/cms-l1t-offline/cmssw/blob/fc3a1a454f40efdf7c95dd97295a6a085a674092/L1Trigger/L1CaloTrigger/plugins/Phase1L1TJetProducer.cc#L215